hyosin-Jang/230425/롤케이크자르기.js

```diff
@@ -0,0 +1,39 @@
+// 롤케이크를 공평하게 자르는 방법의 수 리턴
+// 잘랐을 때 종류의 수가 동일해야 함
+
+function solution(topping) {
+  let [left, right] = [0, topping.length - 1];
+
+  // map에 순회하면서 앞에서부터 센 종류 등장횟수와 뒤에서부터 센 종류 등장 횟수 미리 기록
+  // => O(1) 만에 접근하기 위함
+  let map1 = new Map();
+  let map2 = new Map();
+  let countFromFront = Array.from({length: topping.length}, () => 0);
+  let countFromBack = Array.from({length: topping.length}, () => 0);
+
+  // 앞에서부터 기록
+  topping.forEach((t, idx) => {
+    map1.set(t, (map1.get(t) || 0) + 1);
+    countFromFront[idx] = map1.size;
+  });
+
+  // 뒤에서부터 기록
+  topping.reverse().forEach((t, idx) => {
+    map2.set(t, (map2.get(t) || 0) + 1);
+    countFromBack[idx] = map2.size;
+  });
+
+  countFromBack = countFromBack.reverse();
```

Comment:
```suggestion
  // 뒤에서부터 기록
  let countFromBack = [...countFromFront].reverse();
```
요기서 스프레드 연산으로 받아와도 괜찮을 것 같아요!! 아예 거꾸로 된 맵을 만드는 건 생각도 못했는데..! 아이디어 좋네용

Reply:
오 이게 앞으로부터 세는거랑 뒤에서부터 세는거랑 종류 가짓수가 달라서 countFromFront를 뒤집으면 다른 결과가 나오는거 같아요..!
topping = [1, 0, 2, 1, 3, 4, 1 ,4] 인 경우,
countFromFront = [1, 2, 3, 3, 4, 5, 5, 5]
countFromBack = [5, 5, 4, 3, 3, 2, 2, 1]
이여서 살짝(?) 다릅니다..!

Reply:
헉 그렇네요...! 그 생각을 못했네요 ㅜ